Add an explicit not_after time to the Proof

davidben opened the pull request:
This serves two purposes:

1. It allows a subscriber to easily know when the certificate is expired, without defining a separate certificate property.

2. It allows a subscriber to request a shorter expiration time than the batch would otherwise force. Fixes #83.

In looking at this, I realized that the document's organization is a bit all over the place. Maybe we should just pull the generic BikeshedCertificate parts out into their own section. But for now I'm leaving the overall structure as-is.

davidben commented:
Some notes and philosophizing:

**not_after vs batch expiration time** 

We could either have `not_after` be a constraint on top of what the batch would impose, with the relying party checking the minimum of the two, or say that `not_after` must be consistent with batch expiry, with the relying party rejecting the certificate if inconsistent.

Put another way, is `not_after` the subscriber-requested expiry, or the effective expiry?

I went with effective expiry because this means the subscriber does not need to know anything about the trust anchor to know when the certificate expires. Either it expires at the `not_after` time, or it was simply invalid and will be rejected at any time.

**Proof vs Assertion vs Certificate**

`not_after` could be part of the Assertion, part of the Proof, or just an extra field in the Certificate.

I think the litmus test for being part of the Assertion is "would you put this in a CSR?", and the answer is I think no. "example.com's key is 1234" is a fact about the world. A certificate is the CA making a point-in-time claim that the fact about the world is true.

Now, it being a point-in-time claim is kind of interesting. A CA that issues 14 day certificates and a CA that issues 1 year certificates do not actually perform different validation procedures. It's not like either CA has any better ability to make predictions about the future. So that might suggest that expiration times are all nonsense and maybe we should just say when the certificate was issued, with expiry being a relying party policy?

Except that's not only impractical but also doesn't capture some things. While expiry doesn't tell you anything about the validation process, it tells you how the whole ecosystem will manage this lifecycle of this certificate:

* Until a certificate expires, revocation for that certificate matters
* Until a certificate expires, subscribers should pay attention to it when monitoring transparency logs
* Until a certificate expires, subscribers can reasonably serve it in preference to a newer but larger certificate

Proof vs extra field in the Certificate is kind of meaningless right now. I think the litmus test is, if we end up doing #88, would it be a per-Proof or per-Certificate value? Because this is an effective expiry, I think the answer is per-Proof. Different MTC CAs may force different `not_after` dates, etc.

davidben commented:
Another random thought / justification

**Opaque vs part of framework**

We could either make it part of the BikeshedCertificate framework, or make it something that you put in the specific proof type. Ultimately these are mostly isomorphic. If something is specific to the proof type, subscribers can't read it generically, but then we can define a certificate property that they _do_ know how to read.

But since subscribers should always be able to read this, and these kinds of offline assertions are inherently related to time, putting it in the framework made sense.

bwesterb commented:
Some more things to ponder:

1. How does a subscriber request the earlier expiration? Previously, the client would just send an assertion as assertion-publishing-request. Now we need a `AssertionSigningRequest` different from an Assertion that contains the expiry.

2. Do we need a timestamp or can we use batches are granularity?

3. Maybe also good to ponder revocation again. Depending on how we solve that, we might achieve what we want, by issuing and then immediately revoking for a set time. #41

bwesterb commented:
There is an issue: if you don't store not_after in the abridgedassertion, then the trust service can't compute HashAssertion and thus verify the tree.

davidben commented:
Ah nice catch! I forgot about the HTTP interface.

AbridgedAssertionAndTimestamp is a mouthful. It is annoying that probably every proof type will need a family of "[abridged] assertion and not after" structures and we don't good a word for that. Should it actually go in the assertion?? Trouble is this makes it harder to staple multiple proofs a la #88 if the proofs happen not to be clocked similarly, or have other lifetime constraints, and I kinda like the freedom to build PKIs out of that.

lukevalenta commented on draft-davidben-tls-merkle-tree-certs.md:
```diff
@@ -496,9 +496,9 @@ The CA exposes all of this information in an HTTP {{!RFC9110}} interface describ
 ## Issuance Queue and Scheduling
 
-The CA additionally maintains an issuance queue, not exposed via the HTTP interface.
+The CA additionally maintains an issuance queue, not exposed via the HTTP interface. Each element of the issuance queue is an assertion to be certified, and an optional requested expiration time. The final `not_after` will be computed from both this requested time and the expiration time of the whole batch.
```
In our current implementation (which of course we're happy to change), subscribers submit an `AssertionRequest` to the CA which contains the assertion, evidence, and requested maximum `not_after`. But before queuing that `AssertionRequest`, the CA computes the effective `not_after` as the minimum of the subscriber-requested `not_after`, the next-to-be-issued-batch's expiration time, and any constraints for the evidence (e.g., X.509 cert or revocation check expiry).

The reason we compute the effective `not_after` date at queue time is that otherwise we'd have to re-validate all of the evidence at issuance time, which could be a bottleneck. It also means that we use the next-to-be-issued batch's expiration time instead of the actual batch's expiration time, which prevents assertions from having a longer-than-expected lifetime if there are issuance delays (related to #105).

lukevalenta replied:
Regarding evidence (#23), that's also going into the queue in our implementation as part of the `AssertionRequest`, but should be fine to leave that out of this PR.

davidben replied:
Ah hmm. I mean, both behaviors seem perfectly legitimate to me, so ideally the text would admit both. I guess what you're doing could be modeled as computing a different requested expiration time _before_ entering the queue, which happens to be always less than `batch_not_after`. But that is definitely on the confusing side. Will wordsmith a bit.

davidben commented:
Alright, rebased and then reworked a bunch. This probably warrants a fresh review and possibly looking at individual commits for clarity.

I rebased the PR.

I addressed @lukevalenta's comment by being less prescriptive about where you get the expiration time.

I addressed @bwesterb's comment by putting not_after into AbridgedAssertion.

In doing that, I realized calling this an abridged assertion gets... kinda weird, so I did a bigger change in hopes of resolving this. I tried calling them "batch entries" which seems to work? This resolves a few issues:
* Abridged assertions contain more information than just the assertion (maybe not_after should just be in the assertion??? https://github.com/davidben/merkle-tree-certs/pull/92#issuecomment-2362330928 was the rationale for sticking them with the proof)
* Abridged assertions are currently presented as a funny implementation detail of leaf hashing, but they're more than that. They're also the thing we serve in the HTTP list.
* So the phrasing around "batch entries" tries to talk about them as being *the* contents of the Merkle Tree.

Finally, I noticed a weird inconsistency in how we refer to hash values so cleaned that up.

Thoughts?

bwesterb commented:
> So the phrasing around "batch entries" tries to talk about them as being _the_ contents of the Merkle Tree.

What about *commitments*? Or is that too crypto-y?

lukevalenta reviewed (approved):
Modulo a missing period and whether to call the leaves 'commitments' or 'entries' (I think either is fine, although 'entries' is more succinct), looks good to me!

One implementation consideration (which doesn't need to be part of the spec): right now the [Key()](https://github.com/bwesterb/mtc/blob/main/mtc.go#L1285) function computes a key for an abridged assertion, which is used in various indices and by authenticating parties to request a BikeshedCertificate for a given issued assertion. I think we'll want to omit the `NotAfter` field from that key it's mutable. Instead, the key should be derived from fields in the Assertion structure itself.

bwesterb commented on draft-davidben-tls-merkle-tree-certs.md:
```diff
@@ -475,45 +475,76 @@ The CA also maintains a latest batch number, which is the number of the last bat
+    HashValue subject_info_hash;
+    Claim claims<0..2^16-1>;
+    Timestamp not_after;
+} MerkleTreeEntry;
```
Not ideal to have *batch entry* and *merkle tree entry*. We should make this `BatchEntry` or use *Merkle tree entry* everywhere.

davidben replied:
Ah yeah, done.

bwesterb reviewed (approved):
Don't think naming is ideal yet, but let's just move forward.

davidben commented:
> One implementation consideration (which doesn't need to be part of the spec): right now the [Key()](https://github.com/bwesterb/mtc/blob/main/mtc.go#L1285) function computes a key for an abridged assertion, which is used in various indices and by authenticating parties to request a BikeshedCertificate for a given issued assertion. I think we'll want to omit the NotAfter field from that key as it's mutable. Instead, the key should be derived from fields in the Assertion structure itself.

Ah yeah, another reason to have not_after be part of the Proof rather than the Assertion I guess.

davidben commented:
> What about commitments? Or is that too crypto-y?

I feel like "entry" feels more understandable than "commitment" to me, but I'm perhaps not up on the relevant lingo here. I was just thinking "this is what is actually in the log", because the log does not (necessarily) log the unhashed keys.

*shrug* We can always reshuffle this later.
